### ISSUE TITLE
Errors when using IMPORT DATABASE with the remote protocol

### ISSUE THREAD

**Issue by obi1kenobi** [orientechnologies/orientdb#6758]
## OrientDB Version, operating system, or hardware.
- v2.0 SNAPSHOT[ ] - .18[ ] .17[ ] .16[ ] .15[ ] .14[ ] .13[ ] .12[ ] .11[ ] .10[ ] .9[ ] .8[ ] .7[ ] .6[ ] .5[ ] .4[ ] .3[ ] .2[ ] .1[ ] .0[ ]
- v2.1 SNAPSHOT[ ] - .16[ ] .15[ ] .14[ ] .13[ ] .12[ ] .11[ ] .10[ ] .9[ ] .8[ ] .7[ ] .6[ ] .5[ ] .4[ ] .3[ ] .2[ ] .1[ ] .0[ ]
- v2.2 SNAPSHOT[ ] - .8[x] .rc1[ ] .beta2[ ] .beta1[ ]

## Operating System
- [x] Linux
- [ ] MacOSX
- [ ] Windows
- [ ] Other Unix
- [ ] Other, name? 

## Expected behavior and actual behavior

Expected: `IMPORT DATABASE` successfully imports any file that was previously exported, without regard for what the hardware configuration of the underlying system currently is, or was at the time of the export.

Actual: `IMPORT DATABASE` on a system A using a file exported from system B causes problems if system A has fewer cores than system B. The problems include:
- the console performing `IMPORT DATABASE` hangs
- any subsequent connections to the database being imported also hang (e.g. accessing via Studio)

## Steps to reproduce the problem

On a system with 8 cores (so 8 clusters per class), create about 500 classes, then export the database to a file. Then, on a system with only 1 core and 3GB of RAM (I used a VM, and configured OrientDB with `-Xms512m -Xmx1536m`), pull up the console and run:
```
CREATE DATABASE remote:localhost:2424/test_import <user> <pass> memory graph
IMPORT DATABASE /your_export.json.gz
```

Observe a lot of classes being created like:
```
Creating cluster '<foo>_7'...OK, assigned id=1752
```
and then at one point, observe that the lines of output stop and never resume again. Then try to connect to the database (e.g. via Studio) and attempt to run a simple query, and observe that the "in-progress" spinner keeps spinning and spinning, and the query never returns.

**Comment by lvca** [comment]
This could be a problem of remote import. Could you try by importing the database with plocal?

**Comment by obi1kenobi** [comment]
I regenerated a clean export file from the 8-core machine, when no other processes were using the database. After more testing, I believe this issue may be caused by the same underlying problems as #6687 and #6759.

It is now clear that the database deterministically suffers errors -- the database logs (on both the 1-core and 8-core machines) contain the following stack trace placed in the middle of the import process, which is where the problems start:

```
Error during index rebuild
com.orientechnologies.orient.core.exception.ORecordNotFoundException: The record with id '#128:2' was not found
	DB name="test_import_downscale"
	at com.orientechnologies.orient.core.storage.impl.local.OAbstractPaginatedStorage.readRecord(OAbstractPaginatedStorage.java:1134)
	at com.orientechnologies.orient.core.db.document.ODatabaseDocumentTx$SimpleRecordReader.readRecord(ODatabaseDocumentTx.java:3253)
	at com.orientechnologies.orient.core.db.document.ODatabaseDocumentTx.executeReadRecord(ODatabaseDocumentTx.java:1884)
	at com.orientechnologies.orient.core.tx.OTransactionNoTx.loadRecord(OTransactionNoTx.java:115)
	at com.orientechnologies.orient.core.db.document.ODatabaseDocumentTx.load(ODatabaseDocumentTx.java:1624)
	at com.orientechnologies.orient.core.db.document.ODatabaseDocumentTx.load(ODatabaseDocumentTx.java:103)
	at com.orientechnologies.orient.core.record.impl.ODocument.field(ODocument.java:826)
	at com.orientechnologies.orient.core.index.OPropertyIndexDefinition.getDocumentValueToIndex(OPropertyIndexDefinition.java:72)
	at com.orientechnologies.orient.core.index.OCompositeIndexDefinition.getDocumentValueToIndex(OCompositeIndexDefinition.java:141)
	at com.orientechnologies.orient.core.index.OIndexAbstract.indexCluster(OIndexAbstract.java:994)
	at com.orientechnologies.orient.core.index.OIndexAbstract.fillIndex(OIndexAbstract.java:506)
	at com.orientechnologies.orient.core.index.OIndexAbstract.rebuild(OIndexAbstract.java:467)
	at com.orientechnologies.orient.core.index.OIndexAbstract.rebuild(OIndexAbstract.java:380)
	at com.orientechnologies.orient.core.index.OIndexAbstract.removeCluster(OIndexAbstract.java:709)
	at com.orientechnologies.orient.core.index.OIndexAbstract.removeCluster(OIndexAbstract.java:69)
	at com.orientechnologies.orient.core.index.OIndexManagerAbstract.removeClusterFromIndex(OIndexManagerAbstract.java:205)
	at com.orientechnologies.orient.core.index.OIndexManagerProxy.removeClusterFromIndex(OIndexManagerProxy.java:180)
	at com.orientechnologies.orient.core.metadata.schema.OClassImpl.removeClusterFromIndexes(OClassImpl.java:2685)
	at com.orientechnologies.orient.core.metadata.schema.OClassImpl.removePolymorphicClusterId(OClassImpl.java:2669)
	at com.orientechnologies.orient.core.metadata.schema.OClassImpl.removePolymorphicClusterIds(OClassImpl.java:2656)
	at com.orientechnologies.orient.core.metadata.schema.OClassImpl.removeBaseClassInternal(OClassImpl.java:1284)
	at com.orientechnologies.orient.core.metadata.schema.OSchemaShared.dropClassInternal(OSchemaShared.java:1260)
	at com.orientechnologies.orient.core.metadata.schema.OSchemaShared.dropClass(OSchemaShared.java:530)
	at com.orientechnologies.orient.core.metadata.schema.OSchemaProxy.dropClass(OSchemaProxy.java:146)
	at com.orientechnologies.orient.core.db.tool.ODatabaseImport.removeDefaultNonSecurityClasses(ODatabaseImport.java:661)
	at com.orientechnologies.orient.core.db.tool.ODatabaseImport.importDatabase(ODatabaseImport.java:396)
	at com.orientechnologies.orient.server.network.protocol.binary.ONetworkProtocolBinary.importDatabase(ONetworkProtocolBinary.java:2916)
	at com.orientechnologies.orient.server.network.protocol.binary.ONetworkProtocolBinary.executeRequest(ONetworkProtocolBinary.java:635)
	at com.orientechnologies.orient.server.network.protocol.binary.ONetworkProtocolBinary.sessionRequest(ONetworkProtocolBinary.java:320)
	at com.orientechnologies.orient.server.network.protocol.binary.ONetworkProtocolBinary.execute(ONetworkProtocolBinary.java:196)
	at com.orientechnologies.common.thread.OSoftThread.run(OSoftThread.java:77)
Caused by: java.lang.IllegalArgumentException: Cluster 128 is null
	at com.orientechnologies.orient.core.storage.impl.local.OAbstractPaginatedStorage.getClusterById(OAbstractPaginatedStorage.java:2474)
	at com.orientechnologies.orient.core.storage.impl.local.OAbstractPaginatedStorage.readRecord(OAbstractPaginatedStorage.java:1132)
	... 30 more
$ANSI{green {db=test_import_downscale}} Error executing request
com.orientechnologies.orient.core.db.tool.ODatabaseExportException: Error on importing database 'test_import_downscale' from file: /tmp/test_import_downscaleimport5226924797698865284current_test.json.gz
	at com.orientechnologies.orient.core.db.tool.ODatabaseImport.importDatabase(ODatabaseImport.java:459)
	at com.orientechnologies.orient.server.network.protocol.binary.ONetworkProtocolBinary.importDatabase(ONetworkProtocolBinary.java:2916)
	at com.orientechnologies.orient.server.network.protocol.binary.ONetworkProtocolBinary.executeRequest(ONetworkProtocolBinary.java:635)
	at com.orientechnologies.orient.server.network.protocol.binary.ONetworkProtocolBinary.sessionRequest(ONetworkProtocolBinary.java:320)
	at com.orientechnologies.orient.server.network.protocol.binary.ONetworkProtocolBinary.execute(ONetworkProtocolBinary.java:196)
	at com.orientechnologies.common.thread.OSoftThread.run(OSoftThread.java:77)
Caused by: com.orientechnologies.orient.core.exception.ORecordNotFoundException: The record with id '#128:2' was not found
	DB name="test_import_downscale"
	at com.orientechnologies.orient.core.storage.impl.local.OAbstractPaginatedStorage.readRecord(OAbstractPaginatedStorage.java:1134)
	at com.orientechnologies.orient.core.db.document.ODatabaseDocumentTx$SimpleRecordReader.readRecord(ODatabaseDocumentTx.java:3253)
	at com.orientechnologies.orient.core.db.document.ODatabaseDocumentTx.executeReadRecord(ODatabaseDocumentTx.java:1884)
	at com.orientechnologies.orient.core.tx.OTransactionNoTx.loadRecord(OTransactionNoTx.java:115)
	at com.orientechnologies.orient.core.db.document.ODatabaseDocumentTx.load(ODatabaseDocumentTx.java:1624)
	at com.orientechnologies.orient.core.db.document.ODatabaseDocumentTx.load(ODatabaseDocumentTx.java:103)
	at com.orientechnologies.orient.core.record.impl.ODocument.field(ODocument.java:826)
	at com.orientechnologies.orient.core.index.OPropertyIndexDefinition.getDocumentValueToIndex(OPropertyIndexDefinition.java:72)
	at com.orientechnologies.orient.core.index.OCompositeIndexDefinition.getDocumentValueToIndex(OCompositeIndexDefinition.java:141)
	at com.orientechnologies.orient.core.index.OIndexAbstract.indexCluster(OIndexAbstract.java:994)
	at com.orientechnologies.orient.core.index.OIndexAbstract.fillIndex(OIndexAbstract.java:506)
	at com.orientechnologies.orient.core.index.OIndexAbstract.rebuild(OIndexAbstract.java:467)
	at com.orientechnologies.orient.core.index.OIndexAbstract.rebuild(OIndexAbstract.java:380)
	at com.orientechnologies.orient.core.index.OIndexAbstract.removeCluster(OIndexAbstract.java:709)
	at com.orientechnologies.orient.core.index.OIndexAbstract.removeCluster(OIndexAbstract.java:69)
	at com.orientechnologies.orient.core.index.OIndexManagerAbstract.removeClusterFromIndex(OIndexManagerAbstract.java:205)
	at com.orientechnologies.orient.core.index.OIndexManagerProxy.removeClusterFromIndex(OIndexManagerProxy.java:180)
	at com.orientechnologies.orient.core.metadata.schema.OClassImpl.removeClusterFromIndexes(OClassImpl.java:2685)
	at com.orientechnologies.orient.core.metadata.schema.OClassImpl.removePolymorphicClusterId(OClassImpl.java:2669)
	at com.orientechnologies.orient.core.metadata.schema.OClassImpl.removePolymorphicClusterIds(OClassImpl.java:2656)
	at com.orientechnologies.orient.core.metadata.schema.OClassImpl.removeBaseClassInternal(OClassImpl.java:1284)
	at com.orientechnologies.orient.core.metadata.schema.OSchemaShared.dropClassInternal(OSchemaShared.java:1260)
	at com.orientechnologies.orient.core.metadata.schema.OSchemaShared.dropClass(OSchemaShared.java:530)
	at com.orientechnologies.orient.core.metadata.schema.OSchemaProxy.dropClass(OSchemaProxy.java:146)
	at com.orientechnologies.orient.core.db.tool.ODatabaseImport.removeDefaultNonSecurityClasses(ODatabaseImport.java:661)
	at com.orientechnologies.orient.core.db.tool.ODatabaseImport.importDatabase(ODatabaseImport.java:396)
	... 5 more
Caused by: java.lang.IllegalArgumentException: Cluster 128 is null
	at com.orientechnologies.orient.core.storage.impl.local.OAbstractPaginatedStorage.getClusterById(OAbstractPaginatedStorage.java:2474)
	at com.orientechnologies.orient.core.storage.impl.local.OAbstractPaginatedStorage.readRecord(OAbstractPaginatedStorage.java:1132)
	... 30 more
```

Note that the cluster that causes the error is **always** 128, regardless of when I create the export file. After that error is raised, OrientDB does not terminate the import and give up, but instead raises several more errors, and then finally deletes all data and schema from the database before reporting that the database import completed successfully, as #6759 reported.

Curiously, `IMPORT DATABASE` with `plocal` and the same input file **succeeds** with no errors, and all the data is present.

Given that the problematic cluster ID is a power-of-two, I am tempted to suggest that this is due to a remote import command assuming that IDs will not be larger than one byte in size, and simply failing to import the remaining clusters.

**Comment by obi1kenobi** [comment]
Unfortunately, the file I am using is still full of my company's proprietary information, and I am unable to share it. However, I can now reproduce this error deterministically if you need more information.

**Comment by obi1kenobi** [comment]
@lvca you were right in suggesting that remote import might be at fault. Any luck with debugging and fixing this issue?

**Comment by obi1kenobi** [comment]
I just re-ran my test with 2.2.11, and the following happened:
- None of the above errors were raised, and the console reported `Database import completed in 3045 ms`.
- However, the database was empty -- no data was imported. Furthermore, the console output shows that the import process imported all the schema data, then proceeded to remove all classes, outputting `Removed 260 classes.` just before reporting successful completion.

This means the remote IMPORT DATABASE command is still completely broken.

**Comment by tglman** [comment]
hi @obi1kenobi,

well this is a bit weird, since 2.2.5 if you do a remote import the file is sent to server side and the import is executed server side as embedded, so should not be there much difference between embedded and remote.

About the double display of the log, be aware if you are executing the server.sh and the console.sh in the same terminal you can see double logging of the import, the one done by the server, and the one done by the console.

can you post here the exact steps that you do, just for exclude some command combination that create problem.

If there is nothing that come out of the step the only onther variable is the content of the file.

bye

**Comment by obi1kenobi** [comment]
@tglman thanks for your reply. I am not executing `server.sh` in the same terminal, so the duplication is from the console only.

The export file I use was generated on an 8-core machine, and so has 8 clusters per class. Here are the steps after I've obtained the export file and started the console:
```
orientdb> create database remote:localhost:2424/test_db2 admin admin memory graph

Creating database [remote:localhost:2424/test_db2] using the storage type [memory]...
Connecting to database [remote:localhost:2424/test_db2] with user 'admin'...OK
Database created successfully.

Current database is: remote:localhost:2424/test_db2
orientdb {db=test_db2}> alter database timezone utc                                                  
Database updated successfully.

orientdb {db=test_db2}> import database /current_test.json.gz                                        
Importing database database /current_test.json.gz...
Started import of database 'memory:test_db2' from /tmp/test_db2import5265403511024881116current_test.json.gz...
Non merge mode (-merge=false): removing all default non security classes
- Class E was removed.
- Class V was removed.
- Class ORestricted was removed.
- Class OTriggered was removed.
- Class OSequence was removed.
- Class OSchedule was removed.
- Class OFunction was removed.
Removed 7 classes.
Importing database info...OK
Importing clusters...

<...snip many "Creating cluster" messages...>

Started import of database 'memory:test_db2' from /tmp/test_db2import7887869764074954326current_test.json.gz...
Non merge mode (-merge=false): removing all default non security classes

<...snip many "Class X was removed" messages, where all X are the imported classes...>

Removed 260 classes.
Rebuild of stale indexes...
Start rebuild index ouser.name
Rebuild  of index ouser.name is completed.
Start rebuild index orole.name
Rebuild  of index orole.name is completed.
Stale indexes were rebuilt...
Deleting RID Mapping table...OK


Database import completed in 3045 ms
orientdb {db=test_db2}> exit

<... finally, log into Studio and verify that the database currently has no schema and no data ...>
```

Note that the console prints two "starting import from file X" lines, but the two files X it starts the import from have different names, and their name is altogether different from the original file. Based on this, it sounds like the import is starting twice and interfering with itself when using the remote protocol. Note that this is running on an isolated machine that is not being used by any other process, so this is not a case of me starting two imports by accident.

**Comment by tglman** [comment]
Hi all,

The fix for this was released with 2.2.12, closing this.

bye
